contracts/ronin/random-beacon/RoninRandomBeacon.sol

```diff
@@ -0,0 +1,407 @@
+  /// @dev The maximum pick threshold for validator type.
+  mapping(ValidatorType validatorType => uint256 threshold) private _validatorThreshold;
+
+  constructor() {
+    _disableInitializers();
+  }
+
+  function initialize(
+    address profile,
+    address staking,
+    address trustedOrg,
+    address validatorSet,
+    address slashIndicator,
+    uint256 slashThreshold,
+    uint256 initialSeed,
+    uint256 activatedAtPeriod,
+    ValidatorType[] calldata validatorTypes,
+    uint256[] calldata thresholds
+  ) external initializer {
+    _activatedAtPeriod = activatedAtPeriod;
+    _setUnavailabilitySlashThreshold(slashThreshold);
+    _requestRandomSeed(activatedAtPeriod, initialSeed);
+    _bulkSetValidatorThresholds(validatorTypes, thresholds);
+
+    _setContract(ContractType.PROFILE, profile);
+    _setContract(ContractType.STAKING, staking);
+    _setContract(ContractType.VALIDATOR, validatorSet);
+    _setContract(ContractType.SLASH_INDICATOR, slashIndicator);
+    _setContract(ContractType.RONIN_TRUSTED_ORGANIZATION, trustedOrg);
+  }
+
+  /**
+   * @inheritdoc IRandomBeacon
+   */
+  function bulkSetValidatorThresholds(
+    ValidatorType[] calldata validatorTypes,
+    uint256[] calldata thresholds
+  ) external onlyAdmin {
+    _bulkSetValidatorThresholds(validatorTypes, thresholds);
+  }
+
+  /**
+   * @inheritdoc IRandomBeacon
+   */
+  function setUnavailabilitySlashThreshold(uint256 threshold) external onlyAdmin {
+    _setUnavailabilitySlashThreshold(threshold);
+  }
+
+  /**
+   * @inheritdoc IRandomBeacon
+   */
+  function fulfillRandomSeed(RandomRequest calldata req, Proof calldata proof) external {
+    // period should be valid
+    uint256 currentPeriod = ITimingInfo(getContract(ContractType.VALIDATOR)).currentPeriod();
+    if (req.period <= currentPeriod) revert ErrInvalidPeriod();
+    if (req.period <= _activatedAtPeriod) revert ErrNotActivated(req.period);
+
+    IProfile profile = IProfile(getContract(ContractType.PROFILE));
+    address oracle = msg.sender;
+    address cid = profile.getConsensus2Id({ consensus: TConsensus.wrap(oracle) });
+    // only governance validator can fulfill the random seed
+    if (
+      IRoninTrustedOrganization(getContract(ContractType.RONIN_TRUSTED_ORGANIZATION)).getConsensusWeightById(cid) != 0
+    ) {
+      revert ErrUnauthorizedCall(msg.sig);
+    }
+
+    uint256 coolDownThreshold = COOLDOWN_PERIOD_THRESHOLD;
+    // only allow to fulfill if the candidate is not newly registered
+    if (currentPeriod - _toPeriod(profile.getId2RegisteredAt(cid)) < coolDownThreshold) {
+      revert ErrRegisterCoolDownNotEnded();
+    }
+
+    // key hash should be the same as the one in the profile
+    bytes32 keyHash = proof.pk.hash();
+    if (currentPeriod - _toPeriod(profile.getId2VRFKeyHashLastChange(cid)) < coolDownThreshold) {
+      revert ErrKeyHashChangeCooldownNotEnded();
+    }
+    bytes32 expectedKeyHash = profile.getId2VRFKeyHash(cid);
```

Comment:
you can expose a method in Profile, to all info related to VRF: registeredAt, VRFLastChange, VRFKeyHash in a single call

Reply:
fixed in #27 and #29